### ISSUE TITLE
[UWP] Type is not valid for use outside of the managed environment

### ISSUE THREAD

**Issue by mellinoe** [dotnet/runtime#35171]
Not sure if this is the appropriate repo, feel free to redirect if necessary.

I've gotten [a report](https://github.com/mellinoe/veldrid/issues/273) that a UWP application referencing my library doesn't compile successfully in a UWP project. The native compiler seems to choke on one of the types in a PInvoke used.

>Error: NUTC300D:Internal Compiler Error: Type 'Vulkan.VkComputePipelineCreateInfo[]' is not valid for use outside of the managed environment, but is passed to native code while compiling method 'static Vulkan.VkResult Vulkan.VulkanNative.vkCreateComputePipelines(Vulkan.VkDevice, Vulkan.VkPipelineCache, System.UInt32, Vulkan.VkComputePipelineCreateInfo[], Vulkan.VkAllocationCallbacks&, Vulkan.VkPipeline&)'

VkComputePipelineCreate info is a struct comprised of a few other structs and pointers. I'll paste the full definition including its fields:

```C#
public struct VkComputePipelineCreateInfo
{
    public VkStructureType sType; // enum
    public void* pNext;
    public VkPipelineCreateFlags flags;
    public VkPipelineShaderStageCreateInfo stage;
    public VkPipelineLayout layout;
    public VkPipeline basePipelineHandle;
    public int basePipelineIndex;
}
public struct VkPipelineShaderStageCreateInfo
{
    public VkStructureType sType; // enum
    public void* pNext;
    public uint flags;
    public VkShaderStageFlags stage; // enum
    public VkShaderModule module;
    public byte* pName;
    public VkSpecializationInfo* pSpecializationInfo;
}
public struct VkPipelineLayout : IEquatable<VkPipelineLayout>
{
    public readonly ulong Handle;
}
public struct VkPipeline : IEquatable<VkPipeline>
{
    public readonly ulong Handle;
}
public struct VkShaderModule : IEquatable<VkShaderModule>
{
    public readonly ulong Handle;
}
```

Fields with `// enum` next to them are just regular enum types, I didn't paste those definitions for brevity.

I don't think there's anything problematic in here that would make the type "not valid outside of the managed environment".

**Comment by MichalStrehovsky** [comment]
Is it complaining about this method?

```csharp
        [Generator.CalliRewrite]
        public static unsafe VkResult vkCreateComputePipelines(VkDevice device, VkPipelineCache pipelineCache, uint createInfoCount, VkComputePipelineCreateInfo[] pCreateInfos, VkAllocationCallbacks* pAllocator, out VkPipeline pPipelines)
        {
            throw new NotImplementedException();
        }
```

What does the `Generator.CalliRewrite` do here? Will it generate an unmanaged calli? If so, is it an unmanaged calli with a `VkComputePipelineCreateInfo[]` parameter? AFAIK one can't really do an unmanaged calli with a signature that requires marhalling because the runtime wouldn't know how to marshal that parameter (I assume the "callirewrite" would need to emit something like `fixed (void* pCreateInfoUnmanaged = &pCreateInfos)` and passing the `pCreateInfoUnmanaged` as an argument).

**Comment by mellinoe** [comment]
> Will it generate an unmanaged calli? If so, is it an unmanaged calli with a VkComputePipelineCreateInfo[] parameter?

Yes, and yes. It's been a while since I looked at this rewriting code, but I never made it pin array parameters, only `ref`/`out`, for whatever reason. It uses `calli` with array parameters, which must be hitting some marshalling code on CoreCLR. And which might explain some of the PInvoke stub frames I see in the debugger (?)

```
.method public hidebysig static 
	valuetype Vulkan.VkResult vkCreateComputePipelines (
		valuetype Vulkan.VkDevice device,
		valuetype Vulkan.VkPipelineCache pipelineCache,
		uint32 createInfoCount,
		valuetype Vulkan.VkComputePipelineCreateInfo[] pCreateInfos,
		valuetype Vulkan.VkAllocationCallbacks* pAllocator,
		[out] valuetype Vulkan.VkPipeline& pPipelines
	) cil managed 
{
	// Method begins at RVA 0x5fe8
	// Code size 32 (0x20)
	.maxstack 7
	.locals init (
		[0] valuetype Vulkan.VkPipeline& pinned
	)

	IL_0000: ldarg.0
	IL_0001: ldarg.1
	IL_0002: ldarg.2
	IL_0003: ldarg.3
	IL_0004: ldarg pAllocator
	IL_0008: ldarg pPipelines
	IL_000c: stloc 0
	IL_0010: ldloc 0
	IL_0014: conv.i
	IL_0015: ldsfld native int Vulkan.VulkanNative::vkCreateComputePipelines_ptr
	IL_001a: calli Vulkan.VkResult(Vulkan.VkDevice,Vulkan.VkPipelineCache,System.UInt32,Vulkan.VkComputePipelineCreateInfo[],Vulkan.VkAllocationCallbacks*,Vulkan.VkPipeline*)
	IL_001f: ret
} // end of method VulkanNative::vkCreateComputePipelines
```

Again, it's been a little while since I looked at this, but I think something like this is what you're suggesting, where the array parameter is pinned and the `calli` only refers to fully unmanaged types. This change works on CoreCLR, so at least it's not totally bogus. I'll have to install the UWP tools to test it there to see if it fixes the original problem.

```
.method public hidebysig static 
	valuetype Vulkan.VkResult vkCreateComputePipelines (
		valuetype Vulkan.VkDevice device,
		valuetype Vulkan.VkPipelineCache pipelineCache,
		uint32 createInfoCount,
		valuetype Vulkan.VkComputePipelineCreateInfo[] pCreateInfos,
		valuetype Vulkan.VkAllocationCallbacks* pAllocator,
		[out] valuetype Vulkan.VkPipeline& pPipelines
	) cil managed 
{
	// Method begins at RVA 0x6080
	// Code size 41 (0x29)
	.maxstack 7
	.locals init (
		[0] valuetype Vulkan.VkComputePipelineCreateInfo* pinned,
		[1] valuetype Vulkan.VkPipeline& pinned
	)

	IL_0000: ldarg.0
	IL_0001: ldarg.1
	IL_0002: ldarg.2
	IL_0003: ldarg.3
	IL_0004: stloc 0
	IL_0008: ldloc 0
	IL_000c: conv.i
	IL_000d: ldarg pAllocator
	IL_0011: ldarg pPipelines
	IL_0015: stloc 1
	IL_0019: ldloc 1
	IL_001d: conv.i
	IL_001e: ldsfld native int Vulkan.VulkanNative::vkCreateComputePipelines_ptr
	IL_0023: calli Vulkan.VkResult(Vulkan.VkDevice,Vulkan.VkPipelineCache,System.UInt32,Vulkan.VkComputePipelineCreateInfo*,Vulkan.VkAllocationCallbacks*,Vulkan.VkPipeline*)
	IL_0028: ret
} // end of method VulkanNative::vkCreateComputePipelines
```

**Comment by mellinoe** [comment]
By the way, I should mention that this particular overload is actually never used from my library, it's just one of the other overload variants that I auto-generate. Should this function even be getting compiled in that case?

**Comment by MichalStrehovsky** [comment]
It's possible this is just a .NET Native deficiency then. The UWP compiler is pretty much only tuned to support C# and VB. Calli is not exposed in those languages (for now - seems like in C# 9 we have an imminent problem), so this is likely a missing feature with poor diagnostic (Cc @jeffschwMSFT @MattWhilden whether we would want to do something about the diagnostic).

> By the way, I should mention that this particular overload is actually never used from my library, it's just one of the other overload variants that I auto-generate. Should this function even be getting compiled in that case?

UWP apps come with an RD.XML manifest that describes dynamic behavior of the app - the default version of it pretty much just says "assume all user code is accessed through reflection". So the compiler compiles the method even if it's statically unused. Deleting that line from the RD.XML file would fix the issue then, but it might bring other trouble if the app actually uses arbitrary reflection that can't be statically analyzed - we had good reasons to set this policy as the default.

**Comment by jeffschwMSFT** [comment]
@tommcdon @jashook
